solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -0,0 +1,139 @@
+"""
+
+import pandas as pd
+
+from solarforecastarbiter import datamodel
+from solarforecastarbiter.validation import quality_mapping
+
+
+def apply_validation(data, qfilter, handle_func):
+    """
+    Apply validation steps based on provided filters to the data.
+
+    Parameters
+    ----------
+    data : dict
+        Keys are Observation and Forecast models and values are
+        the timeseries data as pandas.Series.
+    qfilter : solarforecastarbiter.datamodel.QualityFlagFilter
+    handle_func : function
+        Function that handles how `quality_flags` will be used.
+        See solarforecastarbiter.metrics.preprocessing.exclude as an
+        example.
+
+    Returns
+    -------
+    dict :
+        Keys are Observation and Forecast models and values
+        the validated timeseries data as pandas.Series.
+    """
+    validated_data = {}
+
+    # List of flags from filter
+    if not isinstance(qfilter, datamodel.QualityFlagFilter):
+        raise TypeError(f"{qfilter} not a QualityFlagFilter")
+    filters = qfilter.quality_flags
+
+    # Apply handling function to quality flags
+    for model, values in data.items():
+
+        # Apply only to Observations
+        if isinstance(model, datamodel.Observation):
+            if values.empty:
+                validated_data[model] = values.value
+            else:
+                validation_df = quality_mapping.convert_mask_into_dataframe(
+                    values['quality_flag'])
+                validation_df = validation_df[list(filters)]
+                validated_data[model] = handle_func(values.value,
+                                                    validation_df)
+        elif isinstance(model, datamodel.Forecast):
+            validated_data[model] = values
+        else:
+            raise TypeError(f"{model} not an Observation or Forecast")
+
+    return validated_data
+
+
+def resample_and_align(fx_obs, data, tz):
+    """
+    Resample the observation to the forecast interval length and align to
+    remove overlap.
+
+    Parameters
+    ----------
+    fx_obs : solarforecastarbiter.datamodel.ForecastObservation
+        Pair of forecast and observation.
+    data : dict
+        Keys are Observation and Forecast models and values
+        the validated timeseries data as pandas.Series.
+    tz : str
+        Timezone to witch processed data will be converted.
+
+    Returns
+    -------
+    solarforecastarbiter.datamodel.ProcessedForecastObservation
+
+    Todo
+    ----
+      * Add other resampling functions (besides mean like first, last, median)
+    """
+    fx = fx_obs.forecast
+    obs = fx_obs.observation
+
+    # Resample observation
+    closed = datamodel.CLOSED_MAPPING[fx.interval_label]
+    obs_resampled = data[obs].resample(fx.interval_length,
+                                       label=closed,
+                                       closed=closed).mean()
+
+    # Align
+    fx_aligned, obs_aligned = data[fx].align(obs_resampled, 'left')
```

Comment:
If I understand correctly, this line will 1) drop extra points in `obs_resampled` that don't exist in `data[fx]` and 2) add NaNs to `obs_resampled` at times where `data[fx]` exists but `obs_resampled` does not. A comment like that would help. In the long run, we might want to set an observation quality flag here, but that's maybe not needed right now.

Also there's an extra line below.

Reply:
I've added a comment to that effect.